ADR/0028-component-dependencies.md

```diff
@@ -0,0 +1,179 @@
+  * Post the test results back on PR#1 in addition to the results it would normally post to PR#2. It will post the test results back to all PRs in the PR Group.
+
+### Build-service and PR Groups
+
+* When [build-service] responds to the build of a PR (PR #1) and propagates the digest from one Component to another as a PR (PR#2). It follows the declared `depends on` references to know which other repos should receive an update PR. It marks the PR that it submits (PR #2) as being in the same PR Group (slash-prefixed name of the git source branch) as the triggering PR (PR #1). This enables pre-merge testing of both changes.
+* When [build-service] submits PR #2:
+  * It marks it as "Draft" and it includes a reference to the triggering PR (PR #1) in the description of the automatically submitted PR (PR #2), giving the user some indication that it should not be merged out of order.
+  * It checks to see if the two Components (source and destination as determined by the `depends on` references) are in the same git repository (a monorepo) and if PR #1 contains any changes to the `context` directory for the destination Component. If it does, then it create a new commit copying the code changes from that context directory in PR #1 to that context directory in PR #2, and it bases the commit that updates the digest reference on top of this synthetic commit.
+
+## Applied to Use Cases
+
+Let's apply the architecture to some use cases, and see how it plays out:
+
+### Common Parent Image
+
+Scenario: an application image depends on a common parent image. The user has 1 Application, and 2 Components. One of them is the parent image. The other is a child image which uses the parent image in its `FROM` line.
+
+* The child image Component declares that it `depends on` the parent image Component, by way of the new field on the Component CR.
+* [integration-service] will always skip testing for parent image update builds, will never promote them, or use them to initiate Releases, but it will promote them to the global candidate list.
+* [build-service] will propagate digest references as a PR to the child image Component repo, by analyzing the `depends on` fields of all other Components in the Application. The PR that it files must be submitted in such a way that it appears in the same PR Group (the same slash-prefixed name of the git source branch) as the triggering PR submitted by a user.
+* When the parent image PR is merged, [build-service] will update the PRs it originally filed to take them out of "Draft" to indicate that they are safe to merge now as long as there are no other unmerged triggering PRs in the same group. It may potentially rebase the PRs to trigger a new build or use `/retest`.
```

Comment:
What if there are three PR's open:

parent/PR1
parent/PR2
parent/PR3

This will result in three parent PRs:
child/PR1
child/PR2
child/PR3


Each one of those child PRs will reference an child:on-pr-<SHA> build of the parent, and will actively break the child if they are merged (i.e. they should never be merged under any circumstances).

If I then merge all three parent PR's what happens to my three child PRs? Are they all continuously updated to the latest main commit sha (i.e. they become all the same PR)? Is there some logic to close two of them based on the order or merging? 

I have real concerns that this approach will introduce a massive amount of PRs that are never really supposed to be merged, and are not really true pull requests but just a trigger for a PaC run.

This also means that the results of testing are not actually displayed in the PR that is actually mergable, but instead in seperate linked PRs that need to be linked manually.

All this might be fine for toy projects with 1 or 2 PRs a week, but for real projects this is going to get annoying really quickly, and I think will actively harm a projects ability to merge PRs. 

For example just image if we tried to build RHTAP itself using this model, with infra-deployments being a child of every downstream component (build service, image controller, JVM build service etc). It would be impossible to merge anything in infra-deployments, due to the massive volume of PRs that would be created for every downstream PR. 

I would argue that if we want to test the child in response the the parent PR then we should be able to run a pipeline against the parent PR that builds and tests a version of the child and shows the results in the parent PR. This is how it has worked on every project with inter-repo relationships, including RHTAP itself.

Reply:
While there _may_ be a way to partially resolve the issue of the excessive mergable PRs in the child component (see the previous thread), the 1-1 mapping of parent/child PRs on components is a feature for this proposal as it introduces transparency into how the updates are made and what PRs are being linked.

> This also means that the results of testing are not actually displayed in the PR that is actually mergable, but instead in separate linked PRs that need to be linked manually.

This isn't accurate. The proposal includes reporting the test results on all PRs in the dependency tree.

> I would argue that if we want to test the child in response the the parent PR then we should be able to run a pipeline against the parent PR that builds and tests a version of the child and shows the results in the parent PR. This is how it has worked on every project with inter-repo relationships, including RHTAP itself.

This was the premise of #114.